warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java

```diff
@@ -364,7 +367,67 @@ else if (documentRange != null && (!this.isContainsIndexOnlyTerms() && this.getT
             
             pipelineIter.startPipeline();
             
-            this.serializedDocuments = pipelineIter;
+            // gather Key,Document Entries from the pipelines
+            Iterator<Entry<Key,Document>> pipelineDocuments = pipelineIter;
+            
+            if (log.isTraceEnabled()) {
+                pipelineDocuments = Iterators.filter(pipelineDocuments, new Predicate<Entry<Key,Document>>() {
+                    @Override
+                    public boolean apply(@Nullable Entry<Key,Document> keyDocumentEntry) {
+                        log.trace("after pipeline, keyDocumentEntry:" + keyDocumentEntry);
+                        return true;
+                    }
+                });
+            }
+            
+            // now apply the unique transform if requested
+            UniqueTransform uniquify = getUniqueTransform();
+            if (uniquify != null) {
+                pipelineDocuments = Iterators.filter(pipelineDocuments, uniquify.getUniquePredicate());
+            }
+            
+            // apply the grouping transform if requested and if the batch size is greater than zero
+            // if the batch size is 0, then grouping is computed only on the web server
+            GroupingTransform groupify = getGroupingTransform();
+            if (groupify != null && this.groupFieldsBatchSize > 0) {
+                
+                pipelineDocuments = groupingTransform.getGroupingIterator(pipelineDocuments, this.groupFieldsBatchSize);
+                
+                if (log.isTraceEnabled()) {
+                    pipelineDocuments = Iterators.filter(pipelineDocuments, new Predicate<Entry<Key,Document>>() {
+                        @Override
+                        public boolean apply(@Nullable Entry<Key,Document> keyDocumentEntry) {
+                            log.trace("after grouping, keyDocumentEntry:" + keyDocumentEntry);
+                            return true;
+                        }
+                    });
+                }
+                
+            }
+            
+            if (this.getReturnType() == ReturnType.kryo) {
```

Comment:
serialization moved out of the pipeline to here (after unique or grouping is done)